op-plasma/dastate.go

```diff
@@ -0,0 +1,192 @@
+// The input posted onchain is stored in the state for later retrieval.
+func (s *State) SetResolvedChallenge(key []byte, input []byte, resolvedAt uint64) {
+	if c, ok := s.commsByKey[string(key)]; ok {
+		c.challengeStatus = ChallengeResolved
+		c.expiresAt = resolvedAt
+		c.input = input
+		s.metrics.RecordResolvedChallenge(key)
+	}
+}
+
+// SetInputCommitment initializes a new commitment and adds it to the state.
+// This is called when we see a commitment during derivation so we can refer to it later in
+// challenges.
+func (s *State) SetInputCommitment(key []byte, committedAt uint64, challengeWindow uint64) *Commitment {
+	c := &Commitment{
+		key:         key,
+		expiresAt:   committedAt + challengeWindow,
+		blockNumber: committedAt,
+	}
+	s.log.Debug("append commitment", "expiresAt", c.expiresAt, "blockNumber", c.blockNumber)
+	heap.Push(&s.comms, c)
+	s.commsByKey[string(key)] = c
+
+	return c
+}
+
+// GetOrTrackChallenge returns the commitment for the given key if it is already tracked, or
+// initializes a new commitment and adds it to the state.
+func (s *State) GetOrTrackChallenge(key []byte, bn uint64, challengeWindow uint64) *Commitment {
+	if c, ok := s.commsByKey[string(key)]; ok {
+		return c
+	}
+	return s.SetInputCommitment(key, bn, challengeWindow)
+}
+
+// GetResolvedInput returns the input bytes if the commitment was resolved onchain.
+func (s *State) GetResolvedInput(key []byte) ([]byte, error) {
+	if c, ok := s.commsByKey[string(key)]; ok {
+		return c.input, nil
+	}
+	return nil, errors.New("commitment not found")
+}
+
+// ExpireChallenges walks back from the oldest commitment to find the latest l1 origin
+// for which input data can no longer be challenged. It also marks any active challenges
+// as expired based on the new latest l1 origin. If any active challenges are expired
+// it returns an error to signal that a derivation pipeline reset is required.
+func (s *State) ExpireChallenges(bn uint64) (uint64, error) {
+	latest := uint64(0)
+	var err error
+	for i := 0; i < len(s.comms); i++ {
+		c := s.comms[i]
+		if c.expiresAt <= bn && c.blockNumber > latest {
+			latest = c.blockNumber
+
+			if c.challengeStatus == ChallengeActive {
+				c.challengeStatus = ChallengeExpired
+				s.metrics.RecordExpiredChallenge(c.key)
+				err = ErrReorgRequired
+			}
+		} else {
+			break
+		}
+	}
+	return latest, err
+}
+
+// safely prune in case reset is deeper than the finalized l1
+const commPruneMargin = 200
+
+// Prune removes commitments once they can no longer be challenged or resolved.
+func (s *State) Prune(bn uint64) {
+	if bn > commPruneMargin {
+		bn -= commPruneMargin
+	} else {
+		bn = 0
+	}
+	for i := 0; i < len(s.comms); i++ {
+		c := s.comms[i]
+		if c.blockNumber < bn {
+			s.log.Debug("prune commitment", "expiresAt", c.expiresAt, "blockNumber", c.blockNumber)
+			delete(s.commsByKey, string(c.key))
+		} else {
+			// once we're past the given index, remove all commitments
+			s.comms = s.comms[i:]
```

Comment:
Should pop is the above function or re-heapify here here. Truncating the underlying PQ array is not a safe operation.